### PR TITLE
Bug RRT* fix, issues #382 and #383

### PR DESCRIPTION
**Pull request by rafaelrojasmiliani**
#### Reference issue
Fix  issues #382 and #383.

#### What does this implement/fix?
See the issues

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
@rafaelrojasmiliani Hi. Thank you for your PR. Before code review, could you please fix CI error? There are some code style error. Thanks

File: PathPlanning/RRTStar/rrt_star.py [hunk 1]
```diff
@@ -58,28 +60,35 @@ def planning(self, animation=True, search_until_max_iter=True):
...
 
             if self.check_collision(new_node, self.obstacle_list):
                 near_inds = self.find_near_nodes(new_node)
-                new_node = self.choose_parent(new_node, near_inds)
-                if new_node:
+                new_node_2 = self.choose_parent(new_node, near_inds)
+                if new_node_2:
+                    self.rewire(new_node_2, near_inds)
+                    self.node_list.append(new_node_2)
+                else:
```
Comment (AtsushiSakai) [inline PathPlanning/RRTStar/rrt_star.py:85]: This is fixing for #382, correct? If new_node_2 is None, that will be added node_list. I agree to fix it. However, I think `new_node_2` should be `parent_updated_node` or something is better.

File: PathPlanning/RRTStar/rrt_star.py [hunk 2]
```diff
@@ -161,7 +218,11 @@ def rewire(self, new_node, near_inds):
             improved_cost = near_node.cost > edge_node.cost
 
             if no_collision and improved_cost:
-                self.node_list[i] = edge_node
+                near_node.x = edge_node.x
```
Comment (AtsushiSakai) [inline PathPlanning/RRTStar/rrt_star.py:221]: This change is for #383. Do you think just updating edge_node.parent to near_node one works? For example:

edge_node.parent = near_node.parent
self.node_list[i] = edge_node

File: PathPlanning/RRTStar/rrt_star.py [hunk 3]
```diff
@@ -58,28 +60,35 @@ def planning(self, animation=True, search_until_max_iter=True):
...
             print("Iter:", i, ", number of nodes:", len(self.node_list))
             rnd = self.get_random_node()
             nearest_ind = self.get_nearest_node_index(self.node_list, rnd)
-            new_node = self.steer(self.node_list[nearest_ind], rnd, self.expand_dis)
+            new_node = self.steer(
+                self.node_list[nearest_ind], rnd, self.expand_dis)
+            near_node = self.node_list[nearest_ind]
+            new_node.cost = near_node.cost + \
+                math.sqrt((new_node.x-near_node.x)**2 +
+                          (new_node.y-near_node.y)**2)
```
Comment (AtsushiSakai) [inline PathPlanning/RRTStar/rrt_star.py:77]: What is for this change? Related #382?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for your great PR. I have some minior comments and questions. PTAL.

**Comment by rafaelrojasmiliani** [comment]
Hello.
Currently I'm very busy with my job. I will try to work on your suggestion on the weekend.

File: tests/test_rrt_star.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 class Test(TestCase):
 
     def test1(self):
-        m.show_animation = False
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply tests/test_rrt_star.py:20]: no, my bad. Modification removed.

File: tests/test_rrt_star.py [hunk 5]
```diff
@@ -28,7 +28,7 @@ def test_no_obstacle(self):
                              goal=[6, 10],
                              rand_area=[-2, 15],
                              obstacle_list=obstacle_list)
-        path = rrt_star.planning(animation=False)
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply tests/test_rrt_star.py:31]: no, my bad. Modification removed.

File: PathPlanning/RRTStar/rrt_star.py [hunk 6]
```diff
@@ -93,6 +102,21 @@ def planning(self, animation=True, search_until_max_iter=True):
...
+                    randomly generated node with a path from its neared point
+                    There are not coalitions between this node and th tree.
+                near_inds: list
+                    Indices of indices of the nodes what are near to new_node
+
+            Returns.
+            ------
+                Node, a copy of new_node
+        '''
         if not near_inds:
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:120]: Fixed

File: PathPlanning/RRTStar/rrt_star.py [hunk 7]
```diff
@@ -139,17 +165,48 @@ def search_best_goal_node(self):
...
+                new_node: Node
+                    new randomly generated node, without collisions between
+                    its nearest node
+            Returns:
+            -------
+                list
+                    List with the indices of the nodes inside the ball of
+                    radius r
+        '''
         nnode = len(self.node_list) + 1
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:182]: Fixed

File: PathPlanning/RRTStar/rrt_star.py [hunk 8]
```diff
@@ -139,17 +165,48 @@ def search_best_goal_node(self):
...
+        # expand_dist
+        if hasattr(self, 'expand_dis'):
             r = min(r, self.expand_dis)
         dist_list = [(node.x - new_node.x) ** 2 +
                      (node.y - new_node.y) ** 2 for node in self.node_list]
         near_inds = [dist_list.index(i) for i in dist_list if i <= r ** 2]
         return near_inds
 
     def rewire(self, new_node, near_inds):
+        '''
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:194]: Fixed

File: PathPlanning/RRTStar/rrt_star.py [hunk 9]
```diff
@@ -207,9 +269,8 @@ def main():
         if show_animation:
             rrt_star.draw_graph()
             plt.plot([x for (x, y) in path], [y for (x, y) in path], '-r')
-            plt.grid(True)
-            plt.pause(0.01)  # Need for Mac
-            plt.show()
+#            plt.grid(True)
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:272]: done.

File: PathPlanning/RRTStar/rrt_star.py [hunk 10]
```diff
@@ -196,7 +257,8 @@ def main():
                        goal=[6, 10],
                        rand_area=[-2, 15],
                        obstacle_list=obstacle_list)
-    path = rrt_star.planning(animation=show_animation)
+    path = rrt_star.planning(animation=show_animation,
+                             search_until_max_iter=search_until_max_iter)
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:261]: Yes, I'm agree. Now search_until_max_iter is a property of RRTStar.

File: PathPlanning/RRTStar/rrt_star.py [hunk 11]
```diff
@@ -93,6 +102,21 @@ def planning(self, animation=True, search_until_max_iter=True):
         return None
 
     def choose_parent(self, new_node, near_inds):
+        '''
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:105]: Changed to double quotes.

File: PathPlanning/RRTStar/rrt_star.py [hunk 12]
```diff
@@ -93,6 +102,21 @@ def planning(self, animation=True, search_until_max_iter=True):
...
+        Computes the cheapest point to new_node contained in the list
+        near_inds and set such a node as the partent of new_node.
```
[the remark this answers is not in the document]
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:107]: Typo. Fixed.

File: PathPlanning/RRTStar/rrt_star.py (hunk 1 above)
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:85]: `new_node_2` is now `node_with_updated_parent`.

Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:85]: Yes, this should fix #382

File: PathPlanning/RRTStar/rrt_star.py (hunk 2 above)
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:221]: The problem is that there could be a single node such that its parent attribute contains a reference to near_node. The line `self.node_list[i] = edge_node` does not update the children of `self.node_list[i]`. So, if we don't want to look up for the children of `self.node_list[i]` in order to update their parent attribute, we have to update the attributes of instance referenced by its children's parent attribute.

File: PathPlanning/RRTStar/rrt_star.py (hunk 3 above)
Comment (rafaelrojasmiliani) [reply PathPlanning/RRTStar/rrt_star.py:77]: If you remove that line, then the code will start to produce disconnected trees. The problem is that the `cost` attribute was updated inside  `choose_parent`. If `choose_parent` returns `None`, then its value remains undefined.

Comment (AtsushiSakai) [reply PathPlanning/RRTStar/rrt_star.py:78]: OK. I understood. However, I think `np.hypot` should be used here.

File: PathPlanning/RRTStar/rrt_star.py (hunk 2 above)
Comment (AtsushiSakai) [reply PathPlanning/RRTStar/rrt_star.py:226]: OK. Make sense.

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for fixing. I have a last small comment. After fixing it, I will merge it. PTAL.

**Review by AtsushiSakai — APPROVED** [review]
LTGM. Thank you so much for fixing issues : ).

**Pull request merged**